site/app/controllers/ChatroomController.php

```diff
@@ -0,0 +1,234 @@
+<?php
+
+namespace app\controllers;
+
+use app\libraries\Core;
+use app\libraries\response\WebResponse;
+use app\libraries\response\JsonResponse;
+use app\libraries\response\RedirectResponse;
+use app\entities\chat\Chatroom;
+use app\entities\chat\Message;
+use app\libraries\routers\AccessControl;
+use app\libraries\routers\Enabled;
+use app\views\ChatroomView;
+use Symfony\Component\Routing\Annotation\Route;
+use Doctrine\ORM\EntityManagerInterface;
+
+class ChatroomController extends AbstractController {
+
+    public function __construct(Core $core) {
+        parent::__construct($core);
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat", methods={"GET"})
+     */
+    public function showChatroomsPage(): WebResponse {
+        $repo = $this->core->getCourseEntityManager()->getRepository(Chatroom::class);
+        $user = $this->core->getUser();
+        $chatrooms = $repo->findBy([], ['id' => 'ASC']);
+        $active_chatrooms = $repo->findBy(['is_active' => true], ['id' => 'ASC']);
+
+        if ($user->accessAdmin()) {
+            return new WebResponse(
+                'Chatroom',
+                'showChatPageInstructor',
+                $chatrooms
+            );
+        }
+        else { // Student view
+            return new WebResponse(
+                'Chatroom',
+                'showChatPageStudent',
+                $active_chatrooms
+            );
+        }
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/chat/newChatroom", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function addChatroom(): RedirectResponse {
+        $em = $this->core->getCourseEntityManager();
+
+        $chatroom = new Chatroom();
+        $chatroom->setTitle($_POST['title']);
+        $chatroom->setDescription($_POST['description']);
+        $chatroom->setHostId($this->core->getUser()->getId());
+        $chatroom->setHostName($this->core->getUser()->getDisplayFullName());
```

Comment:
Would this be a good opportunity to use a constructor?  See the `Poll` type for an example.